indsl/statistics/outliers.py

```diff
@@ -219,55 +219,66 @@ def _get_outlier_indices(
-
     df["rolling_mean"] = df["val"].rolling(time_window, min_periods=1).mean()  # calculate features
 
     if del_zero_val:  # delete 0 values if user requests it
         df = df[df["val"] != 0.0]
 
     df_dbscan = df[["val"]].copy()
     validate_series_is_not_empty(df_dbscan)
-
-    if eps is None:
+    if eps is None:  # If eps is not given by user, calculate best eps to use for dbscan with nearest neighbors
         # calculate distance to the further neighbor in order to find best epsilon (radius) parameter for dbscan
-        n_neighbors = min(len(df) - 3, 6)
+        n_neighbors = min(max(1, len(df) - 3), 6)
         nbrs = NearestNeighbors(n_neighbors=n_neighbors).fit(df)
         distances, _ = nbrs.kneighbors(df)
         df["knn_dist"] = distances[:, -1]
 
         dist = np.sort(distances[:, -1])
         i = np.arange(len(dist))
-        knee = KneeLocator(i, dist, S=1, curve="convex", direction="increasing", interp_method="polynomial")
-        eps = dist[knee.knee] if dist[knee.knee] > 0 else 0.5  # if calculated eps is 0, use 0.5 as default value
+        try:  # handle cases where np regressor of KneeLocator does not converge
+            knee = KneeLocator(i, dist, S=1, curve="convex", direction="increasing", interp_method="polynomial")
+            if knee.knee is not None:
+                eps = (
+                    dist[knee.knee] if dist[knee.knee] > 0 else 0.5
+                )  # if calculated eps is 0, use 0.5 as default value
+            else:  # there can be cases where KneeLocator converges, but knee.knee is still None
+                eps = 0.5
+        except Exception:  # not catching, as regression might not converge for different reasons
+            eps = 0.5
 
     # train dbscan
     dbscan_model = DBSCAN(eps=eps, min_samples=min_samples).fit(df_dbscan)
     labels = dbscan_model.labels_
 
     outlier_positions = np.where(labels == -1)[0]
-
     # Get indices of outlier data points calculated using dbscan
     outlier_indices_dbscan = df.iloc[outlier_positions].index
 
     # Delete outliers detected by dbscan
     df_without_dbscan_outliers = df.drop(outlier_indices_dbscan)
 
-    # Apply regression on the remaining data points
+    # Apply cubic spline regression on the remaining data points to detect additional outliers
     df_reg = df.loc[df_without_dbscan_outliers.index, :]
-    date_int = df_reg.index.to_series().astype(np.int64)
-    date_int_stand = (date_int - date_int.mean()) / date_int.std()
-    csaps_data = csaps(date_int_stand, df_reg["val"], date_int_stand, smooth=reg_smooth)
-
-    # Delete points with high residuals
-    res = pd.DataFrame(abs(df_reg["val"] - csaps_data))
-    res["val_stand"] = (res["val"] - res["val"].mean()) / res["val"].std()
-    res_stand_outliers = res[res["val_stand"] >= 3]
-
-    outlier_indices_res_std = res_stand_outliers.index
-
-    all_outliers = outlier_indices_dbscan.append(outlier_indices_res_std)
-
+    if (
+        len(df_reg) >= 1
```

Comment:
Performing cubic spline regression on 1 datapoint is not meaningful, so we should think about increasing this threshold. Since it aims to fit the data with a cubic polynomial between each pair of points, having fewer than 4 points might result in an undetermined system where we can't find the coefficients to fit the data